substrate/in-instructions/pallet/src/lib.rs

```diff
@@ -0,0 +1,236 @@
+  #[pallet::getter(fn executed_batches)]
+  pub(crate) type NextBatch<T: Config> = StorageMap<_, Blake2_256, Coin, u64, ValueQuery>;
+
+  #[pallet::hooks]
+  impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+    fn on_finalize(_: BlockNumberFor<T>) {
+      Once::<T>::take();
+    }
+  }
+
+  #[pallet::call]
+  impl<T: Config> Pallet<T> {
+    #[pallet::call_index(0)]
+    #[pallet::weight((0, DispatchClass::Mandatory))] // TODO
+    pub fn execute(origin: OriginFor<T>, updates: Updates) -> DispatchResult {
+      ensure_none(origin)?;
+      assert!(!Once::<T>::exists());
+      Once::<T>::put(true);
+
+      for (coin, update) in updates.iter().enumerate() {
+        if let Some(update) = update {
+          let coin = Coin::from(u32::try_from(coin).unwrap());
+          BlockNumbers::<T>::insert(coin, update.block_number);
+
+          for batch in &update.batches {
+            // TODO: EXECUTE
+            Self::deposit_event(Event::Batch { coin, id: batch.id });
+            NextBatch::<T>::insert(coin, batch.id + 1);
+          }
+        }
+      }
+
+      Ok(())
+    }
+  }
+
+  #[pallet::inherent]
+  impl<T: Config> ProvideInherent for Pallet<T> {
+    type Call = Call<T>;
+    type Error = InherentError;
+    const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
+
+    fn create_inherent(data: &InherentData) -> Option<Self::Call> {
+      data
+        .get_data::<Updates>(&INHERENT_IDENTIFIER)
+        .unwrap()
+        .map(|updates| Call::execute { updates })
+    }
+
+    fn check_inherent(call: &Self::Call, data: &InherentData) -> Result<(), Self::Error> {
+      // Assumes that only not yet handled batches are provided as inherent data
+      if let Some(expected) = data.get_data::<Updates>(&INHERENT_IDENTIFIER).unwrap() {
+        // Match to be exhaustive
+        let updates = match call {
+          Call::execute { ref updates } => updates,
+          _ => Err(InherentError::InvalidCall)?,
+        };
+
+        if updates.len() != expected.len() {
+          Err(InherentError::InvalidUpdateQuantity(
+            updates.len().try_into().unwrap(),
+            expected.len().try_into().unwrap(),
+          ))?;
+        }
+
+        for (coin, both) in updates.iter().zip(&expected).enumerate() {
+          let coin = Coin::from(u32::try_from(coin).unwrap());
+          match both {
+            (Some(update), Some(expected)) => {
+              if update.block_number.0 > expected.block_number.0 {
+                Err(InherentError::UnrecognizedBlockNumber(
+                  coin,
+                  update.block_number,
+                  expected.block_number,
+                ))?;
+              }
+
+              let prev = BlockNumbers::<T>::get(coin);
+              if update.block_number.0 <= prev.0 {
+                Err(InherentError::InvalidBlockNumber(coin, update.block_number, prev))?;
+              }
+
+              if update.batches.len() > expected.batches.len() {
```

Comment:
Based on the way these are populated, I'm failing to see how expected would ever be different than update. Mind elaborating?

Reply:
The former is what the block producer wrote in. The latter is what we would've written in.

A malicious block producer, or one with a different view of the network, may have written in something different.